### PR TITLE
Refactor map size to allow for rectangle maps

### PR DESCRIPTION
**Pull request by IntelOrca**
The map size game action has changed to allow for both width and height to be specified.
This means, plugin API is increased, and any plugins using the map size change action must be updated.
The network version has changed.

Currently the UI has not been changed, so the only way to access rectangle maps right now is to use scripting and invoke the game action with your own width and height.

### CODE REVIEW

File: src/openrct2/actions/ChangeMapSizeAction.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@
 #include "../ui/WindowManager.h"
 #include "../windows/Intent.h"
 
-ChangeMapSizeAction::ChangeMapSizeAction(const int32_t targetSize)
+ChangeMapSizeAction::ChangeMapSizeAction(const TileCoordsXY targetSize)
```
Comment (frutiemax) [inline src/openrct2/actions/ChangeMapSizeAction.cpp:18]: Constant reference.

File: src/openrct2/actions/ChangeMapSizeAction.h [hunk 2]
```diff
@@ -16,7 +16,7 @@ class ChangeMapSizeAction final : public GameActionBase<GameCommand::ChangeMapSi
 {
 public:
     ChangeMapSizeAction() = default;
-    ChangeMapSizeAction(const int32_t targetSize);
+    ChangeMapSizeAction(const TileCoordsXY targetSize);
```
Comment (frutiemax) [inline src/openrct2/actions/ChangeMapSizeAction.h:19]: Constant reference.

File: src/openrct2/interface/Screenshot.cpp [hunk 3]
```diff
@@ -212,10 +212,11 @@ enum class EdgeType
     BOTTOM
 };
 
-static CoordsXY GetEdgeTile(int32_t mapSize, int32_t rotation, EdgeType edgeType, bool visible)
+static CoordsXY GetEdgeTile(TileCoordsXY mapSize, int32_t rotation, EdgeType edgeType, bool visible)
```
Comment (frutiemax) [inline src/openrct2/interface/Screenshot.cpp:215]: Constant reference

File: src/openrct2/world/Map.cpp [hunk 4]
```diff
@@ -1334,9 +1337,9 @@ void map_update_tiles()
         }
 
         // Repeat for each 256x256 block on the map
-        for (int32_t blockY = 0; blockY < gMapSize; blockY += 256)
+        for (int32_t blockY = 0; blockY < gMapSize.y; blockY += 256)
```
Comment (frutiemax) [inline src/openrct2/world/Map.cpp:1340]: Why use signed integer?

File: src/openrct2/world/MapGen.cpp [hunk 5]
```diff
@@ -331,9 +331,9 @@ static void mapgen_place_trees()
     std::vector<TileCoordsXY> availablePositions;
 
     // Create list of available tiles
-    for (int32_t y = 1; y < gMapSize - 1; y++)
+    for (int32_t y = 1; y < gMapSize.y - 1; y++)
```
Comment (frutiemax) [inline src/openrct2/world/MapGen.cpp:334]: Signed integer?

File: src/openrct2/world/MapGen.cpp [hunk 6]
```diff
@@ -409,13 +409,9 @@ static void mapgen_place_trees()
...
 {
-    int32_t x, y, mapSize;
-
-    mapSize = gMapSize;
-
-    for (y = 1; y < mapSize - 1; y++)
+    for (int32_t y = 1; y < gMapSize.y - 1; y++)
     {
-        for (x = 1; x < mapSize - 1; x++)
+        for (int32_t x = 1; x < gMapSize.x - 1; x++)
```
Comment (frutiemax) [inline src/openrct2/world/MapGen.cpp:414]: Signed integer?

File: src/openrct2/world/Map.cpp (hunk 4 above)
Comment (IntelOrca) [reply src/openrct2/world/Map.cpp:1340]: TileCoordsXY is made of signed integer fields, so it it avoids needing any casts.

File: src/openrct2/Context.cpp [hunk 7]
```diff
@@ -502,7 +502,7 @@ namespace OpenRCT2
             viewport_init_all();
 
             _gameState = std::make_unique<GameState>();
-            _gameState->InitAll(150);
+            _gameState->InitAll({ 150, 150 });
```
Comment (Gymnasiast) [inline src/openrct2/Context.cpp:505]: This could use a constant. DEFAULT_MAP_SIZE or something like that.

File: src/openrct2/Editor.cpp [hunk 8]
```diff
@@ -205,9 +205,7 @@ namespace Editor
      */
     static void SetAllLandOwned()
     {
-        int32_t mapSize = gMapSize;
-
-        MapRange range = { 64, 64, (mapSize - 3) * 32, (mapSize - 3) * 32 };
+        MapRange range = { 64, 64, (gMapSize.x - 3) * 32, (gMapSize.y - 3) * 32 };
```
Comment (Gymnasiast) [inline src/openrct2/Editor.cpp:208]:
```suggestion
        MapRange range = { 2 * COORDS_XY_STEP, 2 * COORDS_XY_STEP, (gMapSize.x - 3) * COORDS_XY_STEP, (gMapSize.y - 3) * COORDS_XY_STEP };
```

File: src/openrct2/actions/ChangeMapSizeAction.cpp [hunk 9]
```diff
@@ -33,11 +33,11 @@ void ChangeMapSizeAction::Serialise(DataSerialiser& stream)
 
 GameActions::Result ChangeMapSizeAction::Query() const
 {
-    if (_targetSize > MAXIMUM_MAP_SIZE_TECHNICAL)
+    if (_targetSize.x > MAXIMUM_MAP_SIZE_TECHNICAL || _targetSize.y > MAXIMUM_MAP_SIZE_TECHNICAL)
     {
         return GameActions::Result(GameActions::Status::InvalidParameters, STR_CANT_INCREASE_MAP_SIZE_ANY_FURTHER, STR_NONE);
     }
-    if (_targetSize < 16)
+    if (_targetSize.x < 16 || _targetSize.y < 16)
```
Comment (Gymnasiast) [inline src/openrct2/actions/ChangeMapSizeAction.cpp:40]:
```suggestion
    if (_targetSize.x < MINIMUM_MAP_SIZE_TECHNICAL || _targetSize.y < MINIMUM_MAP_SIZE_TECHNICAL)
```
(This was off-by-one before.)

File: src/openrct2/actions/FootpathPlaceAction.cpp [hunk 10]
```diff
@@ -456,14 +456,15 @@ GameActions::Result FootpathPlaceAction::ElementInsertExecute(GameActions::Resul
  */
 void FootpathPlaceAction::AutomaticallySetPeepSpawn() const
 {
+    auto mapSizeUnits = GetMapSizeUnits();
     uint8_t direction = 0;
     if (_loc.x != 32)
     {
         direction++;
-        if (_loc.y != GetMapSizeUnits() - 32)
+        if (_loc.y != mapSizeUnits.y - 32)
```
Comment (Gymnasiast) [inline src/openrct2/actions/FootpathPlaceAction.cpp:464]:
```suggestion
        if (_loc.y != mapSizeUnits.y - COORDS_XY_STEP)
```

File: src/openrct2/actions/FootpathPlaceAction.cpp [hunk 11]
```diff
@@ -456,14 +456,15 @@ GameActions::Result FootpathPlaceAction::ElementInsertExecute(GameActions::Resul
...
         {
             direction++;
-            if (_loc.x != GetMapSizeUnits() - 32)
+            if (_loc.x != mapSizeUnits.x - 32)
```
Comment (Gymnasiast) [inline src/openrct2/actions/FootpathPlaceAction.cpp:467]:
```suggestion
            if (_loc.x != mapSizeUnits.x - COORDS_XY_STEP)
```

File: src/openrct2/actions/PlaceParkEntranceAction.cpp [hunk 12]
```diff
@@ -51,8 +51,8 @@ GameActions::Result PlaceParkEntranceAction::Query() const
     res.Expenditure = ExpenditureType::LandPurchase;
     res.Position = { _loc.x, _loc.y, _loc.z };
 
-    if (!LocationValid(_loc) || _loc.x <= 32 || _loc.y <= 32 || _loc.x >= (GetMapSizeUnits() - 32)
-        || _loc.y >= (GetMapSizeUnits() - 32))
+    auto mapSizeUnits = GetMapSizeUnits() - CoordsXY{ 32, 32 };
```
Comment (Gymnasiast) [inline src/openrct2/actions/PlaceParkEntranceAction.cpp:54]:
```suggestion
    auto mapSizeUnits = GetMapSizeUnits() - CoordsXY{ COORDS_XY_STEP, COORDS_XY_STEP };
```

File: src/openrct2/actions/PlaceParkEntranceAction.cpp [hunk 13]
```diff
@@ -51,8 +51,8 @@ GameActions::Result PlaceParkEntranceAction::Query() const
...
+    if (!LocationValid(_loc) || _loc.x <= 32 || _loc.y <= 32 || _loc.x >= mapSizeUnits.x || _loc.y >= mapSizeUnits.y)
```
Comment (Gymnasiast) [inline src/openrct2/actions/PlaceParkEntranceAction.cpp:55]:
```suggestion
    if (!LocationValid(_loc) || _loc.x <= COORDS_XY_STEP || _loc.y <= COORDS_XY_STEP || _loc.x >= mapSizeUnits.x || _loc.y >= mapSizeUnits.y)
```

File: src/openrct2/actions/SetCheatAction.cpp [hunk 14]
```diff
@@ -683,12 +683,12 @@ void SetCheatAction::SetStaffSpeed(uint8_t value) const
 
 void SetCheatAction::OwnAllLand() const
 {
-    const int32_t min = 32;
-    const int32_t max = GetMapSizeUnits() - 32;
+    const auto min = CoordsXY{ 32, 32 };
+    const auto max = GetMapSizeUnits() - CoordsXY{ 32, 32 };
```
Comment (Gymnasiast) [inline src/openrct2/actions/SetCheatAction.cpp:687]:
```suggestion
    const auto min = CoordsXY{ COORDS_XY_STEP, COORDS_XY_STEP };
    const auto max = GetMapSizeUnits() - CoordsXY{ COORDS_XY_STEP, COORDS_XY_STEP };
```

File: src/openrct2/cmdline/BenchSpriteSort.cpp [hunk 15]
```diff
@@ -103,8 +102,8 @@ static std::vector<RecordedPaintSession> extract_paint_session(std::string_view
         viewport.var_11 = 0;
         viewport.flags = 0;
 
-        int32_t customX = (gMapSize / 2) * 32 + 16;
-        int32_t customY = (gMapSize / 2) * 32 + 16;
+        int32_t customX = (gMapSize.x / 2) * 32 + 16;
+        int32_t customY = (gMapSize.y / 2) * 32 + 16;
```
Comment (Gymnasiast) [inline src/openrct2/cmdline/BenchSpriteSort.cpp:106]: Note for later: needs to be refactored to use CoordsXY and ToTileCentre()

File: src/openrct2/cmdline/BenchSpriteSort.cpp [hunk 16]
```diff
@@ -87,9 +87,8 @@ static std::vector<RecordedPaintSession> extract_paint_session(std::string_view
         gIntroState = IntroState::None;
         gScreenFlags = SCREEN_FLAGS_PLAYING;
 
-        int32_t mapSize = gMapSize;
-        int32_t resolutionWidth = (mapSize * 32 * 2);
-        int32_t resolutionHeight = (mapSize * 32 * 1);
+        int32_t resolutionWidth = (gMapSize.x * 32 * 2);
+        int32_t resolutionHeight = (gMapSize.y * 32 * 1);
```
Comment (Gymnasiast) [inline src/openrct2/cmdline/BenchSpriteSort.cpp:91]:
```suggestion
        int32_t resolutionWidth = (gMapSize.x * COORDS_XY_STEP * 2);
        int32_t resolutionHeight = (gMapSize.y * COORDS_XY_STEP * 1);
```

File: src/openrct2/interface/Screenshot.cpp [hunk 17]
```diff
@@ -214,10 +214,11 @@ enum class EdgeType
...
 {
+    // TODO
     int32_t lower = (visible ? 1 : 0) * 32;
-    int32_t upper = (visible ? mapSize - 2 : mapSize - 1) * 32;
+    int32_t upper = (visible ? mapSize.x - 2 : mapSize.x - 1) * 32;
```
Comment (Gymnasiast) [inline src/openrct2/interface/Screenshot.cpp:221]:
```suggestion
    int32_t upper = (visible ? mapSize.x - 2 : mapSize.x - 1) * COORDS_XY_STEP;
```

File: src/openrct2/interface/Screenshot.cpp [hunk 18]
```diff
@@ -214,10 +214,11 @@ enum class EdgeType
     BOTTOM
 };
 
-static CoordsXY GetEdgeTile(int32_t mapSize, int32_t rotation, EdgeType edgeType, bool visible)
+static CoordsXY GetEdgeTile(TileCoordsXY mapSize, int32_t rotation, EdgeType edgeType, bool visible)
 {
+    // TODO
     int32_t lower = (visible ? 1 : 0) * 32;
```
Comment (Gymnasiast) [inline src/openrct2/interface/Screenshot.cpp:220]:
```suggestion
    int32_t lower = (visible ? 1 : 0) * COORDS_XY_STEP;
```

File: src/openrct2/interface/Screenshot.cpp [hunk 19]
```diff
@@ -684,11 +685,11 @@ int32_t cmdline_for_screenshot(const char** argv, int32_t argc, ScreenshotOption
...
             }
 
-            int32_t mapSize = gMapSize;
+            const auto& mapSize = gMapSize;
             if (resolutionWidth == 0 || resolutionHeight == 0)
             {
-                resolutionWidth = (mapSize * 32 * 2) >> customZoom;
-                resolutionHeight = (mapSize * 32 * 1) >> customZoom;
+                resolutionWidth = (mapSize.x * 32 * 2) >> customZoom;
+                resolutionHeight = (mapSize.y * 32 * 1) >> customZoom;
```
Comment (Gymnasiast) [inline src/openrct2/interface/Screenshot.cpp:692]:
```suggestion
                resolutionWidth = (mapSize.x * COORDS_XY_STEP * 2) >> customZoom;
                resolutionHeight = (mapSize.y * COORDS_XY_STEP * 1) >> customZoom;
```

File: src/openrct2/Editor.cpp [hunk 20]
```diff
@@ -94,7 +94,7 @@ namespace Editor
     {
         OpenRCT2::Audio::StopAll();
         object_list_load();
-        OpenRCT2::GetContext()->GetGameState()->InitAll(150);
+        OpenRCT2::GetContext()->GetGameState()->InitAll({ 150, 150 });
```
Comment (Gymnasiast) [inline src/openrct2/Editor.cpp:97]:
```suggestion
        OpenRCT2::GetContext()->GetGameState()->InitAll({ DEFAULT_MAP_SIZE, DEFAULT_MAP_SIZE });
```

File: src/openrct2/Editor.cpp [hunk 21]
```diff
@@ -170,7 +170,7 @@ namespace Editor
 
         object_manager_unload_all_objects();
         object_list_load();
-        OpenRCT2::GetContext()->GetGameState()->InitAll(150);
+        OpenRCT2::GetContext()->GetGameState()->InitAll({ 150, 150 });
```
Comment (Gymnasiast) [inline src/openrct2/Editor.cpp:173]:
```suggestion
        OpenRCT2::GetContext()->GetGameState()->InitAll({ DEFAULT_MAP_SIZE, DEFAULT_MAP_SIZE });
```

File: src/openrct2/Editor.cpp [hunk 22]
```diff
@@ -191,7 +191,7 @@ namespace Editor
...
```
Comment (Gymnasiast) [inline src/openrct2/Editor.cpp:194]:
```suggestion
        OpenRCT2::GetContext()->GetGameState()->InitAll({ DEFAULT_MAP_SIZE, DEFAULT_MAP_SIZE });
```

File: src/openrct2/interface/Screenshot.cpp [hunk 23]
```diff
@@ -701,9 +702,9 @@ int32_t cmdline_for_screenshot(const char** argv, int32_t argc, ScreenshotOption
             if (customLocation)
             {
                 if (centreMapX)
-                    customX = (mapSize / 2) * 32 + 16;
```
Comment (Gymnasiast) [inline src/openrct2/interface/Screenshot.cpp:704]: customX/customY should be refactored in another PR (making a note so that I don’t forget).

File: src/openrct2/park/ParkFile.cpp [hunk 24]
```diff
@@ -891,8 +891,8 @@ namespace OpenRCT2
             auto found = os.ReadWriteChunk(
                 ParkFileChunkType::TILES,
                 [pathToSurfaceMap, pathToQueueSurfaceMap, pathToRailingsMap](OrcaStream::ChunkStream& cs) {
-                    cs.ReadWrite(gMapSize); // x
-                    cs.Write(gMapSize);     // y
```
Comment (Gymnasiast) [inline src/openrct2/park/ParkFile.cpp:895]: Nice one, planning this ahead :+1:

File: src/openrct2/ride/TrackDesign.cpp [hunk 25]
```diff
@@ -1993,10 +1993,10 @@ static bool TrackDesignPlacePreview(TrackDesignState& tds, TrackDesign* td6, mon
     uint8_t backup_rotation = _currentTrackPieceDirection;
     uint32_t backup_park_flags = gParkFlags;
     gParkFlags &= ~PARK_FLAGS_FORBID_HIGH_CONSTRUCTION;
-    int32_t mapSize = gMapSize << 4;
+    auto mapSize = TileCoordsXY{ gMapSize.x / 16, gMapSize.y / 16 };
```
Comment (Gymnasiast) [inline src/openrct2/ride/TrackDesign.cpp:1996]: This is reversed. `<< 4` corresponds with `* 16`, not `/ 16`.

File: src/openrct2/ride/TrackDesign.cpp [hunk 26]
```diff
@@ -2148,7 +2148,7 @@ static void TrackDesignPreviewClearMap()
 {
     auto numTiles = MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL;
 
-    gMapSize = 256;
+    gMapSize = { 256, 256 };
```
Comment (Gymnasiast) [inline src/openrct2/ride/TrackDesign.cpp:2151]: This _used_ to be the maximum map size. Since it’s ludicrously big for a track design preview already, I propose creating a constant for it (e.g. TRACK_DESIGN_PREVIEW_MAP_SIZE).

File: src/openrct2/title/TitleScreen.cpp [hunk 27]
```diff
@@ -129,7 +129,7 @@ void TitleScreen::Load()
     GetContext()->GetNetwork().Close();
 #endif
     OpenRCT2::Audio::StopAll();
-    GetContext()->GetGameState()->InitAll(150);
+    GetContext()->GetGameState()->InitAll({ 150, 150 });
```
Comment (Gymnasiast) [inline src/openrct2/title/TitleScreen.cpp:132]:
```suggestion
    GetContext()->GetGameState()->InitAll({ DEFAULT_MAP_SIZE, DEFAULT_MAP_SIZE });
```

File: src/openrct2/title/TitleScreen.cpp [hunk 28]
```diff
@@ -335,7 +335,7 @@ bool TitleScreen::TryLoadSequence(bool loadPreview)
         _loadedTitleSequenceId = SIZE_MAX;
         if (!loadPreview)
         {
-            GetContext()->GetGameState()->InitAll(150);
+            GetContext()->GetGameState()->InitAll({ 150, 150 });
```
Comment (Gymnasiast) [inline src/openrct2/title/TitleScreen.cpp:338]:
```suggestion
            GetContext()->GetGameState()->InitAll({ DEFAULT_MAP_SIZE, DEFAULT_MAP_SIZE });
```

File: src/openrct2/world/Map.cpp [hunk 29]
```diff
@@ -2304,3 +2307,14 @@ void FixLandOwnershipTilesWithOwnership(std::initializer_list<TileCoordsXY> tile
         }
     }
 }
+
+MapRange ClampRangeWithinMap(const MapRange& range)
+{
+    auto mapSizeMax = GetMapSizeMaxXY();
+    auto aX = std::max<decltype(range.GetLeft())>(32, range.GetLeft());
```
Comment (Gymnasiast) [inline src/openrct2/world/Map.cpp:2314]:
```suggestion
    auto aX = std::max<decltype(range.GetLeft())>(COORDS_XY_STEP, range.GetLeft());
```

File: src/openrct2/world/Map.cpp [hunk 30]
```diff
@@ -2304,3 +2307,14 @@ void FixLandOwnershipTilesWithOwnership(std::initializer_list<TileCoordsXY> tile
...
+    auto bX = std::min<decltype(range.GetRight())>(mapSizeMax.x, range.GetRight());
+    auto aY = std::max<decltype(range.GetTop())>(32, range.GetTop());
```
Comment (Gymnasiast) [inline src/openrct2/world/Map.cpp:2316]:
```suggestion
    auto aY = std::max<decltype(range.GetTop())>(COORDS_XY_STEP, range.GetTop());
```

File: src/openrct2/actions/PlacePeepSpawnAction.cpp [hunk 31]
```diff
@@ -46,8 +46,9 @@ GameActions::Result PlacePeepSpawnAction::Query() const
     res.Expenditure = ExpenditureType::LandPurchase;
     res.Position = _location;
 
-    if (!LocationValid(_location) || _location.x <= 16 || _location.y <= 16 || _location.x >= (GetMapSizeUnits() - 16)
-        || _location.y >= (GetMapSizeUnits() - 16))
+    auto mapSizeUnits = GetMapSizeUnits() - CoordsXY{ 16, 16 };
```
Comment (Gymnasiast) [inline src/openrct2/actions/PlacePeepSpawnAction.cpp:49]: We should have a constant for 16 too, I think. Something like `HALF_TILE`. But this can be done in another PR too.

**Comment by Gymnasiast** [comment]
Getting there, found 1 possible mistake and a few code style/quality things.

File: src/openrct2/ride/TrackDesign.cpp (hunk 25 above)
Comment (IntelOrca) [reply src/openrct2/ride/TrackDesign.cpp:1996]: Well spotted!

File: src/openrct2/interface/Screenshot.cpp [hunk 32]
```diff
@@ -214,10 +214,11 @@ enum class EdgeType
     BOTTOM
 };
 
-static CoordsXY GetEdgeTile(int32_t mapSize, int32_t rotation, EdgeType edgeType, bool visible)
+static CoordsXY GetEdgeTile(TileCoordsXY mapSize, int32_t rotation, EdgeType edgeType, bool visible)
 {
-    int32_t lower = (visible ? 1 : 0) * 32;
-    int32_t upper = (visible ? mapSize - 2 : mapSize - 1) * 32;
+    // TODO
```
Comment (Gymnasiast) [inline src/openrct2/interface/Screenshot.cpp:219]: It’s not clear what this TODO entails, so could you add a description?

**Review by Gymnasiast — APPROVED** [review]
